Libraries/LibIPC/TransportSocket.cpp

```diff
@@ -13,26 +13,76 @@
 
 namespace IPC {
 
+void SendQueue::enqueue_message(Vector<u8>&& bytes, Vector<int>&& fds)
+{
+    Threading::MutexLocker locker(m_mutex);
+    m_bytes.append(bytes.data(), bytes.size());
+    m_fds.append(fds.data(), fds.size());
+    m_condition.signal();
+}
+
+SendQueue::Running SendQueue::block_until_message_enqueued()
+{
+    Threading::MutexLocker locker(m_mutex);
+    while (m_bytes.is_empty() && m_fds.is_empty() && m_running)
+        m_condition.wait();
+    return m_running ? Running::Yes : Running::No;
+}
+
+SendQueue::BytesAndFds SendQueue::dequeue(size_t max_bytes)
+{
+    Threading::MutexLocker locker(m_mutex);
+    auto bytes_to_send = min(max_bytes, m_bytes.size());
+    Vector<u8> bytes;
+    bytes.append(m_bytes.data(), bytes_to_send);
+    m_bytes.remove(0, bytes_to_send);
+    return { move(bytes), move(m_fds) };
+}
+
+void SendQueue::return_unsent_data_to_front_of_queue(ReadonlyBytes const& bytes, Vector<int> const& fds)
+{
+    Threading::MutexLocker locker(m_mutex);
+    m_bytes.prepend(bytes.data(), bytes.size());
+    m_fds.prepend(fds.data(), fds.size());
+}
+
+void SendQueue::stop()
+{
+    Threading::MutexLocker locker(m_mutex);
+    m_running = false;
+    m_condition.signal();
+}
+
 TransportSocket::TransportSocket(NonnullOwnPtr<Core::LocalSocket> socket)
     : m_socket(move(socket))
 {
     m_send_queue = adopt_ref(*new SendQueue);
     m_send_thread = Threading::Thread::construct([this, send_queue = m_send_queue]() -> intptr_t {
         for (;;) {
-            send_queue->mutex.lock();
-            while (send_queue->messages.is_empty() && send_queue->running)
-                send_queue->condition.wait();
-
-            if (!send_queue->running) {
-                send_queue->mutex.unlock();
+            if (send_queue->block_until_message_enqueued() == SendQueue::Running::No)
                 break;
-            }
 
-            auto [bytes, fds] = send_queue->messages.take_first();
-            send_queue->mutex.unlock();
+            auto [bytes, fds] = send_queue->dequeue(4096);
+            ReadonlyBytes remaining_to_send_bytes = bytes;
 
-            if (auto result = send_message(*m_socket, bytes, fds); result.is_error()) {
+            auto result = send_message(*m_socket, remaining_to_send_bytes, fds);
+            if (result.is_error()) {
                 dbgln("TransportSocket::send_thread: {}", result.error());
+                VERIFY_NOT_REACHED();
+            }
+
+            if (!remaining_to_send_bytes.is_empty() || !fds.is_empty()) {
+                send_queue->return_unsent_data_to_front_of_queue(remaining_to_send_bytes, fds);
+            }
+
+            if (m_socket->fd().has_value()) {
```

Comment:
this check was missing

Reply:
Could this be written instead as

```
if (!is_open())
   return; // continue?
```

I suspect that if someone has shut down this TransportSocket, we probably want to end the thread asap?

Reply:
actually, is there a data race here? if someone closes the fd, or assigns it to zero, how is that handled in a way that doesn't cause a data race with the main thread and the send thread here?

Reply:
> actually, is there a data race here?

yes, since sending thread was a part of IPC::Connection we were pretending like it's fine to close it while read/write on another thread. I made another commit that adds RWLock for underlying socket, so we never try to read/write while close() is invoked.

> Could this be written instead as

sure, applied this change.